pwndbg/heap/ptmalloc.py

```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                    tcache = tcache_addr
+
+            try:
+                self._thread_cache = pwndbg.memory.poi(self.tcache_perthread_struct, tcache)
+                _ = self._thread_cache['entries'].fetch_lazy()
+            except Exception as e:
+                print(message.error('Error fetching tcache. GDB cannot access '
+                                    'thread-local variables unless you compile with -lpthread.'))
+                return None
+
+            return self._thread_cache
+
+        print(message.warn('This version of GLIBC was not compiled with tcache support.'))
+        return None
+
+    @property
+    def mp(self):
+        self._mp_addr = pwndbg.symbol.address('mp_')
+        if self._mp_addr is not None:
+            self._mp = pwndbg.memory.poi(self.malloc_par, self._mp_addr)
+
+        return self._mp
+
+    @property
+    def global_max_fast(self):
+        self._global_max_fast_addr = pwndbg.symbol.address('global_max_fast')
+        if self._global_max_fast_addr is not None:
+            self._global_max_fast = pwndbg.memory.u(self._global_max_fast_addr)
+
+        return self._global_max_fast
+
+    @property
+    @pwndbg.memoize.reset_on_objfile
+    def heap_info(self):
+        return pwndbg.typeinfo.load('heap_info')
+
+    @property
+    @pwndbg.memoize.reset_on_objfile
+    def malloc_chunk(self):
+        return pwndbg.typeinfo.load('struct malloc_chunk')
+
+    @property
+    @pwndbg.memoize.reset_on_objfile
+    def malloc_state(self):
+        return pwndbg.typeinfo.load('struct malloc_state')
+
+    @property
+    @pwndbg.memoize.reset_on_objfile
+    def tcache_perthread_struct(self):
+        return pwndbg.typeinfo.load('struct tcache_perthread_struct')
+
+    @property
+    @pwndbg.memoize.reset_on_objfile
+    def tcache_entry(self):
+        return pwndbg.typeinfo.load('struct tcache_entry')
+
+    @property
+    @pwndbg.memoize.reset_on_objfile
+    def mallinfo(self):
+        return pwndbg.typeinfo.load('struct mallinfo')
+
+    @property
+    @pwndbg.memoize.reset_on_objfile
+    def malloc_par(self):
+        return pwndbg.typeinfo.load('struct malloc_par')
+
+
+    def get_heap(self, addr):
+        """Find & read the heap_info struct belonging to the chunk at 'addr'."""
+        return pwndbg.memory.poi(self.heap_info, heap_for_ptr(addr))
+
+    def get_arena(self, arena_addr=None):
+        """Read a malloc_state struct from the specified address, default to
+        reading the current thread's arena. Return the main arena if the
+        current thread is not attached to an arena.
+        """
+        if arena_addr is None:
+            if self.multithreaded:
+                arena_addr = pwndbg.memory.u(pwndbg.symbol.address('thread_arena'))
+                if arena_addr > 0:
+                    return pwndbg.memory.poi(self.malloc_state, arena_addr)
+
+            return self.main_arena
+
+        try:
+            next(i for i in pwndbg.vmmap.get() if arena_addr in i)
```

Comment:
@lebr0nli What's the point of this `next(...)`?

Reply:
It want to make sure the arena address is a valid address in memory (when `next` failed, it means we can't find it in memory)?
But it seems a little bit expensive because it use exception to handle it, maybe I should use `pwndbg.vmmap.find(arena_addr) is None` to determine it?
I'm not sure, I just refactor some of the original methods from `Heap` to `DebugSymsHeap` 😆